common/ManagementTestShared/ManagementRecordTestBase.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+using Azure.Core.TestFramework;
+using Azure.ResourceManager.Resources;
+using System.Diagnostics;
+
+using System.Threading.Tasks;
+using NUnit.Framework;
+
+
+namespace Azure.ResourceManager.TestFramework
+{
+    [Category("Recorded")]
+    public abstract class ManagementRecordTestBase<TEnvironment> : ClientTestBase where TEnvironment : TestEnvironment, new()
+    {
+        protected RecordedTestSanitizer Sanitizer { get; set; }
+
+        protected RecordMatcher Matcher { get; set; }
+
+        public TEnvironment TestEnvironment { get; }
+
+        private ResourceGroupCleanupPolicy CleanupPolicy { get; set; }
+
+        protected ManagementRecordTestBase(bool isAsync) : this(isAsync, RecordedTestUtilities.GetModeFromEnvironment())
+        {
+        }
+        protected ManagementRecordTestBase(bool isAsync, RecordedTestMode mode) : base(isAsync)
+        {
+            Sanitizer = new RecordedTestSanitizer();
+            Matcher = new RecordMatcher();
+            Mode = mode;
+            TestEnvironment = new TEnvironment();
+            TestEnvironment.Mode = Mode;
+            CleanupPolicy = new ResourceGroupCleanupPolicy();
+        }
+
+        /// <summary>
+        /// Initializes the clients for a test scenario so the recordings are setup and saved
+        /// correctly
+        /// </summary>
+        protected abstract void InitializeClients();
+
+        /// <summary>
+        /// Setup of resources that will be run before a test fixture or suite
+        /// This should be overriden if resources need to be setup once for a suite of test
+        /// </summary>
+        protected virtual Task OnOneTimeSetupAsync()
```

Comment:
Hmmm, this is tricky. The user may not really know a [OneTimeSetup] ran depending on how familiar they are with NUnit. So I think this could confuse a user into thinking they have to make a one time setup and this will run after it. 

OnOneTimeSetup I guess could also be misleading to someone knowing NUnit more as well.

Maybe some other naming here could mitigate both??? Maybe the documentation will be sufficient and we just name it Afterxxx

Reply:
I think if we were going to have a BeforeSetupAsync() then this would be necessary.  As it is, I think OnSetupAsync is more familiar as a pattern, and the idea is they really shouldn't have to worry about when the base class code is running.  I would argue against having Before and After, as I think we want to simplify what the dev has to do as much as possible.

Reply:
That is precisely why the context of After is more useful.  It is letting the consumer know that you are doing some one time set up and whatever they add will run After that.

This is a very common pattern in .NET where On typically overrides behavior, After / Before adds behavior.  We aren't overriding any behavior here.